---
output: github_document
---

<!-- README.md is generated from README.Rmd. Please edit that file -->

```{r, include = FALSE}
options(width = 400)
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "man/figures/README-",
  out.width = "100%",
  results = "hold"
)
```

# ass

<!-- badges: start -->
<!-- badges: end -->

The goal of ass is to simulate RNA-seq reads with alternative splicing events.

## Installation

<!-- You can install the released version of ass from [CRAN](https://CRAN.R-project.org) with: -->

<!-- ``` r -->
<!-- install.packages("ass") -->
<!-- ``` -->

You can install the development version from [GitHub](https://github.com/) with:

``` r
# install.packages("remotes")
Sys.setenv(GITHUB_PAT = "16946cf3c89e74cd07363d0f660559e4339d730b")
remotes::install_github("biomedbigdata/as_simulator")
```

Please note that we use a custon Version of Polyester, that is available at https://github.com/quirinmanz/polyester

## Example

### Creating exon supersets

```{r createSuperset}
suppressMessages(library(ass))

# create exon superset for genes on chromosome 21 of ensembl release 99
gtf_file = system.file('extdata', 'Homo_sapiens.GRCh38.99.21.gtf', package = 'ass')
# by default the produed superset will be saved as .rda file into the same directory
exon_superset = get_exon_supersets(gtf_file)
exon_superset[[1]][1:5, ]
```

### Simulating Alternative Splicing

You can find more information about the main function of this package at the end of the page.

```{r simulateASS}
# now we create splice variants from 9 exon supersets
max_genes = 9

# to create one variant per gene we set prob_as_freq to TRUE and then produce one of each AS events as well as one variant containing every event
probs_as_freq = T
event_freq = 
    setNames(rep(1/9, 9),
             c('es', 'mes', 'ir', 'a3', 'a5', 'afe', 'ale', 'mee', 'es,ir,mes,a3,a5,afe,ale,mee'))

# we will create a small experiment with just one samplesper group
num_reps = c(1,1)

# we can use the previously created superset to simulate splice variants from, since it is saved in the same directory as the gtf
# if no superset is found, a new one will be created
simulate_alternative_splicing(input_dir = system.file('extdata', package = 'ass'),
                              event_probs = event_freq,
                              outdir = 'simulation', 
                              probs_as_freq = probs_as_freq, 
                              max_genes = max_genes,
                              num_reps = num_reps)
```


<!-- What is special about using `README.Rmd` instead of just `README.md`? You can include R chunks like so: -->

<!-- ```{r cars} -->
<!-- #summary(cars) -->
<!-- ``` -->

<!-- You'll still need to render `README.Rmd` regularly, to keep `README.md` up-to-date. -->

<!-- You can also embed plots, for example: -->

### Visualize Splice Variants

```{r visualization, dpi=300, results = "hold"}
# to visualitze the splice variants we will use ggbio
suppressMessages(library(ggbio))

# firstly, we load the newly created gtf file 
gtf = rtracklayer::import('simulation/splicing_variants.gtf')

# the gene id of the variant with all events
gene_id = gtf$gene_id[grep('es,ir,mes,a3,a5,afe,ale,mee', gtf$transcript_id, fixed = T)[1]]
exons = gtf[gtf$type == 'exon' & gtf$gene_id == gene_id]
suppressWarnings(ggbio::autoplot(split(exons, exons$transcript_id)))

# check annotation
event_anno = read.csv('simulation/event_annotation.tsv', sep = '\t')
event_anno[grepl(gene_id, event_anno$template) | grepl(gene_id, event_anno$variant), ]
```



## `simulate_alternative_splicing`: Simulate RNA-seq experiment with splicing variants

### Description


 Firstly, exon supersets are created by joining all exons of a gene from a gtf file.
 Next, splicing variants are created with documentation and event annotation based on the users input.
 Finally, fastq files containing RNA-seq reads from the splice variants and the real exon and junction coverage
 are created using a modified version of the polyester R package available on https://github.com/quirinmanz/polyester.


### Usage

```r
simulate_alternative_splicing(input_dir, event_probs, outdir, ncores = 1L, ...)
```


### Arguments

Argument      |Description
------------- |----------------
```input_dir```     |     Character path to directory containing the gtf file from which splice variants are created and genome fasta files passed to polyester.
```event_probs```     |     Named list/vector containing numerics corresponding to the probabilites to create the event(combination). If `probs_as_freq` is `TRUE`  `event_probs` correspond to the relative frequency of occurences for the event(combination) and in this case the sum of all frequencies has to be <=1.
```outdir```     |     character, path to folder where simulated reads and all annotations should be written, with *no* slash at the end. By default, reads are written to current working directory.
```ncores```     |     the number of cores to be utilized for parallel generation of splice variant creation and read simulation.
```...```     |     any of several other arguments that can be used to add nuance to the simulation and splice variant creation. See details.

### Details


 Reads are simulated from a GTF file which is pruduced by `create_splicing_variants_and_annotation` plus DNA sequences.
 
 Several optional parameters can be passed to this function to adjust the simulation. For polyester parameters refer to `simulate_experiment` from the polyester R package:
 
    

*   `write_gff` : Additionally to the gtf file containing the splice variants, a gff3 file with the same content will be printed to the outdir. Default `TRUE`   

*   `max_genes` : The maximum number of genes/exon supersets to be included in the process of splice variant creation. Default `NULL` which means that all available exon supersets will be used.  

*   `exon_junction_coverage` : Should the real coverage of exons, junctions and retained introns should be written into a additional file. Default `TRUE`   

*   `multi_events_per_exon` : Should it be possible to have more than one AS event at the same exon if multiple variants are created for the same exon superset? !If this option is set to `TRUE` , there may occur unforeseen AS events that are not documented in the event_annotation file!. Default `FALSE`   

*   `probs_as_freq` : Should `event_probs` be treated as relative frequencies instead of probabilities? Default `FALSE`   

*   `save_exon_superset` : Should the exon supersets be saved to .rda file? Default `TRUE`   
 
 Parameters passed to polyester that we assigned different defaults to than in `simulate_experiment`:

*   `fold_changes` : Currently, ass introduces random isform switches. Those can be retraced in the sim_tx_info.txt file written by polyester. We plan on improving this in the future.  

*   `strand_specific` : Strand-specific simulation (1st read forward strand, 2nd read reverse strand with respect to transcript sequence). Default `TRUE` .  

*   `meanmodel` : `reads_per_transcripts` as a function of transcript length. Default `TRUE` .  

*   `verbose` : Should progress messages be printed during the sequencing process? Default `TRUE` .  

*   `exon_junction_coverage` : Should the coverage of exons, junctions and retained introns be determined? Default `TRUE` .  

*   `exon_junction_table` : If `exon_junction_coverage=TRUE` a `data.table` produced by `create_splicing_variants_and_annotation`  to determine exon and intron coverage.  


### Value


 No return, but simulated reads, a simulation info file, an alternative splicing event annotation and exon and junction coverages are written to `outdir` .


### References


 Alyssa C. Frazee, Andrew E. Jaffe, Ben Langmead, Jeffrey T. Leek,
 Polyester: simulating RNA-seq datasets with differential transcript expression,
 Bioinformatics, Volume 31, Issue 17, 1 September 2015, Pages 2778â€“2784,
 https://doi.org/10.1093/bioinformatics/btv272



<!-- In that case, don't forget to commit and push the resulting figure files, so they display on GitHub! -->
