---
output: github_document
---

```{r, include = FALSE}
options(width = 400)
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "man/figures/README-",
  out.width = "100%",
  results = "hold"
)
```

# ASimulatoR

The goal of ASimulatoR is to simulate RNA-seq reads with alternative splicing events. The alternative splicing events are well documented and the true origin of each read is used for exon and juntion coverage via a modified version of the bioconductor polyester package.

## Installation

You can install the development version from [GitHub](https://github.com/) with:

``` r
# install.packages("remotes")
remotes::install_github("biomedbigdata/ASimulatoR")
```

Please note that we use a custom version of Polyester, that is available at https://github.com/biomedbigdata/polyester

## Example

### Creating exon supersets
Firstly, we create exon supersets by joining all exons of a gene from a gtf/gff file. 
These supersets are then used to create splice variants. 
Since all exons from one gene are used to create the exon superset, you may find that the term exon superset is used analogously to gene.

```{r createSuperset}
suppressMessages(library(ASimulatoR))

# create exon superset for genes on chromosome 21 of ensembl release 99
gtf_file = system.file('extdata', 'Homo_sapiens.GRCh38.99.21.gtf', package = 'ASimulatoR')

# by default the produced superset will be saved as .rda file into the same directory
exon_superset = get_exon_supersets(gtf_file)
exon_superset[[1]][1:5, ]
```

### Simulating Alternative Splicing

You can find more information about the main function of this package at the end of the page.

This simulator supports eight different AS events:

| es           | mes                    | ir               | a3                                  |  a5                              | mee                      | afe                    | ale                   |
|--------------|------------------------|------------------|-------------------------------------|----------------------------------|--------------------------|------------------------|-----------------------|
| exon skiping | multiple exon skipping | intron retention | alternative 3'/acceptor splice site | alternative 5'/donor splice site | mutually exclusive exons | alternative first exon | alternative last exon |


```{r parameters}
# define your input_dir, where the annotation gtf (or the exon supersets if you have already created them) and the genome fasta files are located
# here we will use the example data
input_dir = system.file('extdata', package = 'ASimulatoR')

# define, how many groups and samples per group you analyze. Here we create a small experiment with two groups with one sample per group:
num_reps = c(1,1)

# define your outdir with NO slash
outdir = 'simulation'

# define the number of genes you want to work with. If you want all exons, do not specify this parameter or set it to NULL
# here we create splice variants from 9 exon supersets:
max_genes = 9
```

You could define the distribution of the events by probability or relative frequency. 

* Probability: For each superset we create an event with the probability mentioned in `event_prob`.
* Frequency: Set `probs_as_freq = T`. The exon supersets are partitioned corresponding to the `event_prob` parameter.


```{r simulateAS}
# in this example we use relative frequencies
# here we produce eight variants with one of each AS events as well as one variant containing every event
# if probs_as_freq was FALSE, a random number would be drawn for each event-superset combination and only if it was smaller than 1/9 the AS event would be created
probs_as_freq = T
event_freq = 
    setNames(rep(1/9, 9),
             c('es', 'mes', 'ir', 'a3', 'a5', 'afe', 'ale', 'mee', 'es,ir,mes,a3,a5,afe,ale,mee'))



# we use the previously created superset to simulate splice variants from, since it is saved in the same directory as the gtf
# if no superset is found, a new one will be created
simulate_alternative_splicing(input_dir = input_dir,
                              event_probs = event_freq,
                              outdir = outdir, 
                              probs_as_freq = probs_as_freq, 
                              max_genes = max_genes,
                              num_reps = num_reps)
```


### Visualize Splice Variants

```{r visualization, dpi=300, results = "hold"}
# to visualize the splice variants we will use ggbio
suppressMessages(library(ggbio))

# firstly, we load the newly created gtf file 
gtf = rtracklayer::import('simulation/splicing_variants.gtf')

# the gene id of the variant with all events
gene_id = gtf$gene_id[grep('es,ir,mes,a3,a5,afe,ale,mee', gtf$transcript_id, fixed = T)[1]]
exons = gtf[gtf$type == 'exon' & gtf$gene_id == gene_id]
suppressWarnings(ggbio::autoplot(split(exons, exons$transcript_id)))

# have a look at the event annotation
event_anno = read.csv('simulation/event_annotation.tsv', sep = '\t')
event_anno[grepl(gene_id, event_anno$template) | grepl(gene_id, event_anno$variant), ]
```



## `simulate_alternative_splicing`: Simulate RNA-seq experiment with splicing variants

### Description


 Firstly, exon supersets are created by joining all exons of a gene from a gtf/gff file.
 Next, splicing variants are created with documentation and event annotation based on the users input.
 Finally, fastq files containing RNA-seq reads from the splice variants and the real exon and junction coverage
 are created using a modified version of the polyester R package available on https://github.com/biomedbigdata/polyester.


### Usage

```r
simulate_alternative_splicing(input_dir, event_probs, outdir, ncores = 1L, ...)
```


### Arguments

Argument      |Description
------------- |----------------
```input_dir```     |     Character path to directory containing the gtf/gff file from which splice variants are created and genome fasta files with one file per chromosome i.e. <chr_name>.fa passed to polyester
```outdir```     |     character, path to folder where simulated reads and all annotations should be written, with *no* slash at the end. By default, reads are written to current working directory.
```event_probs```     |     Named list/vector containing numerics corresponding to the probabilites to create the event (combination). If `probs_as_freq` is `TRUE`  `event_probs` correspond to the relative frequency of occurences for the event(combination) and in this case the sum of all frequencies has to be <=1.
```ncores```     |     the number of cores to be utilized for parallel generation of splice variant creation and read simulation.
```preset```     |     if you want to use preset parameters one of 'event_partition', 'rna_seq_experiment', 'event_combination_2'. Check `?presets` for more information
```...```     |     any of several other arguments that can be used to add nuance to the simulation and splice variant creation. See details.

### Details


 Reads are simulated from a GTF file which is produced by `create_splicing_variants_and_annotation` plus DNA sequences.
 
 Several optional parameters can be passed to this function to adjust the simulation. For polyester parameters refer to `simulate_experiment` from the polyester R package:
 
    
*   `novel_variants` : Numeric value between 0 and 1 indicating the percentage of splicing variants that will be suppressed in an additional gtf file splicing_variants_novel.gtf.

*   `write_gff` : Additionally to the gtf file containing the splice variants, a gff3 file with the same content will be printed to the outdir. Default `TRUE`   

*   `max_genes` : The maximum number of genes/exon supersets to be included in the process of splice variant creation. Default `NULL` which means that all available exon supersets will be used.  

*   `exon_junction_coverage` : Should the real coverage of exons, junctions and retained introns be written into a additional file. Default `TRUE`   

*   `multi_events_per_exon` : Should it be possible to have more than one AS event at the same exon if multiple variants are created for the same exon superset? !If this option is set to `TRUE` , there may occur unforeseen AS events that are not documented in the event_annotation file!. Default `FALSE`   

*   `probs_as_freq` : Should `event_probs` be treated as relative frequencies instead of probabilities? Default `FALSE`   

*   `save_exon_superset` : Should the exon supersets be saved to .rda file? Default `TRUE`   
 
 Parameters passed to polyester that we assigned different defaults to than in `simulate_experiment`:

*   `fold_changes` : Currently, ASimulatoR introduces random isoform switches. Those can be retraced in the sim_tx_info.txt file written by polyester. We plan on improving this in the future.  

*   `strand_specific` : Strand-specific simulation (1st read forward strand, 2nd read reverse strand with respect to transcript sequence). Default `TRUE` .  

*   `meanmodel` : `reads_per_transcripts` as a function of transcript length.  Always `TRUE` in ASimulatoR.  

*   `frag_GC_bias`: A sample-specific GC content bias on the fragment level. Currently not supported in ASimulatoR: always 'none'.

*   `verbose` : Should progress messages be printed during the sequencing process? Default `TRUE` .  

*   `exon_junction_coverage` : Should the coverage of exons, junctions and retained introns be determined? Default `TRUE` .  

*   `exon_junction_table` : If `exon_junction_coverage=TRUE` a `data.table` produced by `create_splicing_variants_and_annotation`  to determine exon and intron coverage.  


### Value


 No return, but simulated reads, a simulation info file, an alternative splicing event annotation and exon and junction coverages are written to `outdir` .


### References


 Alyssa C. Frazee, Andrew E. Jaffe, Ben Langmead, Jeffrey T. Leek,
 Polyester: simulating RNA-seq datasets with differential transcript expression,
 Bioinformatics, Volume 31, Issue 17, 1 September 2015, Pages 2778â€“2784,
 https://doi.org/10.1093/bioinformatics/btv272

### License


ASimulatoR Copyright (C) 2020  Manz, Quirin

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <https://www.gnu.org/licenses/>.